notebooks/6.2. Validating Network - Google Directions API.ipynb

```diff
@@ -461,6 +757,193 @@
     "n.edge('9791490', '4698712638')"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3.2 Validating google speed values"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Once you have attached the google speed values to the network, you may want to do some validation, to check if the values make sense and if there are any missing values. \n",
+    "\n",
+    "To do that, we first need to convert the network to a geodataframe."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# If you only sent the requests for parts of the network with a certain OSM tag, \n",
+    "# you should pass the list of those tags to the function graph_to_gdf()\n",
+    "\n",
+    "def graph_to_gdf(network, osm_tag=all):\n",
+    "    subgraph_t = network.subgraph_on_link_conditions(\n",
+    "        conditions = [\n",
+    "            {'attributes': {'osm:way:highway': {'text': osm_tag}}},\n",
+    "            {'modes' : 'car'}],\n",
+    "        how = all, \n",
+    "        mixed_dtypes = True)\n",
+    "    \n",
+    "    # convert subgraph to geodataframe\n",
+    "    gdf_dict = generate_geodataframes(subgraph_t)\n",
+    "    gdf = gdf_dict['links']\n",
+    "    \n",
+    "    # fill in missing points (due to network structure when filtering by osm_tag)  \n",
+    "    gdf.loc[gdf['geometry'].isna(), 'geometry'] = gdf.loc[gdf['geometry'].isna()].apply(\n",
+    "        lambda x: line_geometry(x['from'], x['to']), axis=1)\n",
+    "    \n",
+    "    # convert to epsg:4326 to allow visualisation in Kepler in section 4\n",
+    "    gdf = gdf.to_crs(\"epsg:4326\")\n",
+    "    \n",
+    "    return gdf\n",
+    "\n",
+    "\n",
+    "# For filling in missing points\n",
+    "def line_geometry(u, v):\n",
+    "    from_node = n.node(u)\n",
+    "    to_node = n.node(v)\n",
+    "    return LineString(\n",
+    "        [(float(from_node['x']), float(from_node['y'])), (float(to_node['x']), float(to_node['y']))])"
```

Comment:
So I wonder if we need this. Just looked up the geodataframe method and it looks to have this included: https://github.com/arup-group/genet/blob/master/genet/outputs_handler/geojson.py#L30 I don't remember why we needed this before and I wonder, is it that something in this method is not working correctly? we use it for a lot of other things so would be good to check this out